py/server/deephaven/_wrapper.py

```diff
@@ -162,15 +163,35 @@ def pythonify(j_obj: Any) -> Optional[Any]:
     return wrap_j_object(j_obj)
 
 
+def _wrap_with_subclass(j_obj: jpy.JType, cls: type) -> Optional[JObjectWrapper]:
+    """ Returns a wrapper instance for the specified Java object if the specified class is a subclass of the wrapper
+    class. Otherwise, returns None. """
+    subclasses = cls.__subclasses__()
+    for subclass in subclasses:
+        try:
+            if (wrapper := _wrap_with_subclass(j_obj, subclass)) is not None:
+                return wrapper
+            return subclass(j_obj)
+        except:
+            continue
+    return None
+
 def wrap_j_object(j_obj: jpy.JType) -> Union[JObjectWrapper, jpy.JType]:
-    """ Wraps the specified Java object as an instance of a custom wrapper class if one is available, otherwise returns
-    the raw Java object. """
+    """ Wraps the specified Java object as an instance of the most specific custom wrapper class if one is available,
+    otherwise returns the raw Java object. """
     if j_obj is None:
         return None
 
-    wc = _lookup_wrapped_class(j_obj)
+    wcs = _lookup_wrapped_class(j_obj)
+    for wc in wcs:
```

Comment:
The order of `wcs` seems potentially arbitrary -- or there needs to be some documentation about why it isn't.  If it is arbitrary, there is a problem.  I have concerns that if it isn't arbitrary there are still problems.

See the note above that I do not think this matches what Ryan and I agreed to.

Reply:
See the modified code example above, I think the order being arbitrary isn't a problem as long as all the classes are distinguishable from each other.

Reply:
There is a hidden contract about uniqueness of constructors.  I think that should get documented here, since I don't think it is documented elsewhere.

Reply:
It is sorta documented, see last bullet. 
```python
def _wrap_with_subclass(j_obj: jpy.JType, cls: type) -> Optional[JObjectWrapper]:
    """ Returns a wrapper instance for the specified Java object by trying the entire subclasses' hierarchy. The
    function employs a Depth First Search strategy to try the most specific subclass first. If no matching wrapper class is found,
    returns None.

    The premises for this function are as follows:
    - The subclasses all share the same class attribute `j_object_type` (guaranteed by subclassing JObjectWrapper)
    - The subclasses are all direct initialisable (guaranteed by subclassing JObjectWrapper)
    - The subclasses are all distinct from each other and check for their uniqueness in the initializer (__init__), e.g.
      InputTable checks for the presence of the INPUT_TABLE_ATTRIBUTE attribute on the Java object.
    """
```

Reply:
OK